<a href="https://colab.research.google.com/github/mthonorio/IA/blob/main/prova_1/IA_Prova_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1º Prova de Inteligência Artificial 2021.2
#### **Aluno:** Matheus B. Honório
#### **Matrícula:** 20190097098

## Questão 1:

**(1 Ponto)** Para cada um dos agentes a seguir, desenvolva uma descrição completando a tabela abaixo sobre o ambiente de tarefas de cada um. 

(a) Agente comentarista de jogo. 

(b) Agente sugestor de vinho.

(c) Agente treinador de time (como de basquete).

*Obs:* Para o preenchimento da tabela abaixo foi escolhido como esporte dos itens **a** e **c** o basquete.

| Agente | Medida de desempenho | Ambiente | Atuadores | Sensores |
| :----: | :------------------ | :------- | :-------- | :------- |
| a | Deve falar sobre a pontuação e movimento dos jogadores <br/> e das torcidas | Quadra, Times, Pontuação e Arbitragem | Microfone e Caixas de Som | Câmeras, Placar e Movimento |
| b | Notificar o cliente sobre o melhor vinho para o evento proposto,<br/> suas principais características e preços | Restaurantes, Vinícolas, Vinhos e <br/> Eventos de amostra | Monitor e Caixa de Som | Câmera, Odômetro e Microfone |
| c | Maximizar as estratégias e treinamento do time por completo e <br/> individualmente, imprimir exercícios e estatísticas dos jogadores | Quadra e Conjunto de jogadores | Monitor, Impressora e Caixa de Som | Câmera, Odômetro e Microfone |
 

## Questão 2:

**(2 Pontos)** Diferencie os 4 tipos de aprendizado de máquina. Para tanto, escolha uma base de dados do Kaggle e explique como aplicar os 4 métodos nessa mesma base.

A base escolhida para a questão foi a [MovieLens 20M Dataset](https://www.kaggle.com/datasets/grouplens/movielens-20m-dataset).

### Aprendizagem supervisionada


O algoritmo de aprendizado (indutor) recebe um conjunto de exemplos de treinamento no qual os rótulos das classes são conhecidos para construir um classificador que possa determinar corretamente a classe de novos exemplos ainda não rotulados. Portanto, aplicando o método de aprendizagem supervisionada na nossa base de dados abordando como objetivo principal um modelo de **classificação** de gêneros de filmes de forma que ao submeter uma descrição e seu título é possível obter os seus gêneros.

### Aprendizagem não supervisionada

O método de aprendizagem não supervisionada busca analisar e induzir os exemplos fornecidos determinando se alguns deles podem ser agrupados de alguma maneira, formando agrupamentos ou clusters, ou seja, em nossa base de dados ao submetermos os filmes da base de dados, o algoritmo separará os filmes e agrupá-los em gêneros ou outra categoria. 

### Aprendizagem semi-supervisionada

O método de aprendizagem semi-supervisionada é o treinamento de classificadores quando uma grande de exemplos não rotulados está disponível acompanhado de um pequeno conjunto de exemplos rotulados. Portanto, em nossa base de dados poucos dados são conhecidos e uma grande parte desconhecida, podemos utilizar o aprendizado para classificar, por exemplo, classificação dos filmes utilizando semelhanças observáveis pela descrição, título ou direção.

### Aprendizagem por reforço

A aprendizagem por reforço trata de situações onde um agente aprende por tentativa e erro ao atuar sobre um ambiente dinâmico, ou seja, no nosso caso através da tentativa e erro a IA pode rotular a classificação etária do filme através de seu gênero devido ao diretor do título, descrição e gênero.

## Questão 3:

**(5 Pontos)** Em um Jupyter Notebook (link do collab, GitHub ou arquivo), utilize o método K vizinhos mais próximos na base escolhida na questão anterior para um contexto SUPERVISIONADO e para classificação:

1. Aplique processos de pré-processamento e os necessários para execução do método K-NN.
2. Aplique correlação ou PCA e selecione um número menor de atributos com esses métodos e justifique sua decisão.
3. Escolha 5 valores de K e mostre os resultados (escolha 3 métricas de avaliação e mostre a matriz de confusão) e explique os resultados encontrados.
4. Escolha 2 métricas de similaridade (diferentes distâncias ou outras métricas - correlação, por exemplo) para definir o(s) vizinho(s) mais próximos e explique os resultados encontrados.
5. Melhor K e melhor distância e testar com o conjunto de atributos do item 2.


In [40]:
#bibliotecas necessárias para ler/manipular/visualizar dados
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#bibliotecas para criar o modelo de ML
from scipy.sparse import csr_matrix
from sklearn.neighbors import NearestNeighbors
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.decomposition import PCA
from sklearn import metrics
from sklearn.metrics import confusion_matrix

In [41]:
# Lendo a base de dados e atribuindo a uma variável
data_movie = pd.read_csv("movie.csv")
data_rating = pd.read_csv("rating.csv")

In [33]:
# Acessando o grupo de colunas e atribuindo a variável
movie = data_movie.loc[:,{"movieId","title"}]
rating = data_rating.loc[:,{"userId","movieId","rating"}]

In [34]:
# Juntando(Mergeando) nossas bases de dados em uma única variável
data = pd.merge(movie,rating)
# Indexando nossa base 
data = data.iloc[:1000000,:]
# Criando uma tabela dinâmica como um DataFrame.
user_movie_table = data.pivot_table(index = ["title"],columns = ["userId"],values = "rating").fillna(0)
user_movie_table.head(10)

userId                                 1       2       3       4       5       \
title                                                                           
Ace Ventura: When Nature Calls (1995)     0.0     0.0     0.0     3.0     0.0   
Across the Sea of Time (1995)             0.0     0.0     0.0     0.0     0.0   
Amazing Panda Adventure, The (1995)       0.0     0.0     0.0     0.0     0.0   
American President, The (1995)            0.0     0.0     0.0     0.0     5.0   
Angela (1995)                             0.0     0.0     0.0     0.0     0.0   
Angels and Insects (1995)                 0.0     0.0     0.0     0.0     0.0   
Anne Frank Remembered (1995)              0.0     0.0     0.0     0.0     0.0   
Antonia's Line (Antonia) (1995)           0.0     0.0     0.0     0.0     0.0   
Assassins (1995)                          0.0     0.0     0.0     0.0     0.0   
Babe (1995)                               0.0     0.0     0.0     0.0     0.0   

userId                                 6       7       8       10      11      \
title                                                                           
Ace Ventura: When Nature Calls (1995)     0.0     0.0     1.0     0.0     3.5   
Across the Sea of Time (1995)             0.0     0.0     0.0     0.0     0.0   
Amazing Panda Adventure, The (1995)       0.0     0.0     0.0     0.0     0.0   
American President, The (1995)            0.0     4.0     0.0     4.0     0.0   
Angela (1995)                             0.0     0.0     0.0     0.0     0.0   
Angels and Insects (1995)                 0.0     0.0     0.0     0.0     0.0   
Anne Frank Remembered (1995)              0.0     0.0     0.0     0.0     0.0   
Antonia's Line (Antonia) (1995)           0.0     0.0     0.0     0.0     0.0   
Assassins (1995)                          0.0     0.0     0.0     0.0     0.0   
Babe (1995)                               0.0     0.0     0.0     0.0     0.0   

userId                                 ...  138484  138485  138486  138487  \
title                                  ...                                   
Ace Ventura: When Nature Calls (1995)  ...     2.0     0.0     0.0     0.0   
Across the Sea of Time (1995)          ...     0.0     0.0     0.0     0.0   
Amazing Panda Adventure, The (1995)    ...     0.0     0.0     0.0     0.0   
American President, The (1995)         ...     4.0     0.0     0.0     0.0   
Angela (1995)                          ...     0.0     0.0     0.0     0.0   
Angels and Insects (1995)              ...     0.0     0.0     0.0     0.0   
Anne Frank Remembered (1995)           ...     0.0     0.0     0.0     0.0   
Antonia's Line (Antonia) (1995)        ...     0.0     0.0     0.0     0.0   
Assassins (1995)                       ...     4.0     0.0     0.0     0.0   
Babe (1995)                            ...     0.0     0.0     3.5     0.0   

userId                                 138488  138489  138490  138491  138492  \
title                                                                           
Ace Ventura: When Nature Calls (1995)     0.0     0.0     0.0     0.0     0.0   
Across the Sea of Time (1995)             0.0     0.0     0.0     0.0     0.0   
Amazing Panda Adventure, The (1995)       0.0     0.0     0.0     0.0     0.0   
American President, The (1995)            0.0     0.0     0.0     0.0     0.0   
Angela (1995)                             0.0     0.0     0.0     0.0     0.0   
Angels and Insects (1995)                 0.0     0.0     4.0     0.0     0.0   
Anne Frank Remembered (1995)              0.0     0.0     0.0     0.0     0.0   
Antonia's Line (Antonia) (1995)           0.0     0.0     3.0     0.0     0.0   
Assassins (1995)                          0.0     0.0     0.0     0.0     0.0   
Babe (1995)                               0.0     0.0     5.0     0.0     0.0   

userId                                 138493  
title                                          
Ace Ventura: When Nature Calls (1995)     4.0  


In [42]:
data.head()

movieId             title  rating  userId
0        1  Toy Story (1995)     4.0       3
1        1  Toy Story (1995)     5.0       6
2        1  Toy Story (1995)     4.0       8
3        1  Toy Story (1995)     4.0      10
4        1  Toy Story (1995)     4.5      11

In [12]:
# Escolhendo um filme aleatoriamente
query_index = np.random.choice(user_movie_table.shape[0])
print("O filme selecionado é: ",user_movie_table.index[query_index])

O filme selecionado é:  Unforgettable (1996)


In [46]:
# Criando uma matriz de valores a partir da tabela
user_movie_table_matrix = csr_matrix(user_movie_table.values)
# Utilizando a métrica do coseno para a classificação e algoritmo automático
model_knn = NearestNeighbors(metric = 'cosine', algorithm = 'auto')
# Iniciando o algoritmo de treinamento
model_knn.fit(user_movie_table_matrix)
# Atribuindo as variáveis valores do modelo e buscando pelos 6 vizinhos mais próximos
distances, indices = model_knn.kneighbors(user_movie_table.iloc[query_index,:].values.reshape(1,-1), n_neighbors = 6)

In [56]:
# Criando a lista vazia
movie = []
distance = []

for i in range(0, len(distances.flatten())):
    if i != 0:
        movie.append(user_movie_table.index[indices.flatten()[i]])
        distance.append(distances.flatten()[i])    

m=pd.Series(movie,name='movie')
d=pd.Series(distance,name='distance')
recommend = pd.concat([m,d], axis=1)
recommend = recommend.sort_values('distance',ascending=False)

# Printa os 5 vizinhos mais próximos
print('Recomendações para {0}:\n'.format(user_movie_table.index[query_index]))
for i in range(0,recommend.shape[0]):
    print('{0}: {1}, com a distância de {2}'.format(i, recommend["movie"].iloc[i], recommend["distance"].iloc[i]))

Recomendações para Unforgettable (1996):

0: Mary Reilly (1996), com a distância de 0.8183690340547402
1: Before and After (1996), com a distância de 0.8140652234565855
2: Juror, The (1996), com a distância de 0.8089523799680194
3: City Hall (1996), com a distância de 0.8044178662544587
4: Eye for an Eye (1996), com a distância de 0.7781912617093903


## Questão 4:

**(2 Pontos)** Considerando a matriz de confusão abaixo, quais são os valores das métricas listadas abaixo?

Obs: **Classe verdadeira - Coluna | Classe predita - Linha**

| **CLASSES** | Maçã | Laranja | Limão | Manga |
| ------- | :--: | :-----: | :---: | :---: |
| Maçã | 70 | 20 | 5 | 5 |
| Laranja | 10 | 67 | 3 | 20 | 
| Limão | 5 | 30 | 55 | 10 | 
| Manga | 5 | 4 | 2 | 1000 |

Maçã:

| **CLASSES** | Maçã | Laranja | Limão | Manga |
| ------- | :--: | :-----: | :---: | :---: | 
| Maçã | VP | FP | FP | FP |
| Laranja | FN | VN | VN | VN |
| Limão | FN | VN | VN | VN |
| Manga | FN | VN | VN | VN |

Laranja:

| **CLASSES** | Maçã | Laranja | Limão | Manga |
| ------- | :--: | :-----: | :---: | :---: | 
| Maçã | VN | FN | VN | VN |
| Laranja | FP | VP | FP | FP |
| Limão | VN | FN | VN | VN |
| Manga | VN | FN | VN | VN |

Limão:

| **CLASSES** | Maçã | Laranja | Limão | Manga |
| ------- | :--: | :-----: | :---: | :---: | 
| Maçã | VN | VN | FN | VN |
| Laranja | VN | VN| FN | VN |
| Limão | FP | FP | VP | FP |
| Manga | VN | VN | FN | VN |

Manga:

| **CLASSES** | Maçã | Laranja | Limão | Manga |
| ------- | :--: | :-----: | :---: | :---: | 
| Maçã | VN | VN | VN | FN |
| Laranja | VN | VN | VN | FN |
| Limão | VN | VN | VN | FN |
| Manga | FP | FP | FP | VP |

Organizando nossos valores em uma tabela, temos:

| **CLASSES** | VP | VN | FP | FN |
| ------- | :--: | :-----: | :---: | :---: | 
| Maçã | 70 | 1191 | 30 | 20 |
| Laranja | 67 | 1157 | 33 | 54 | 
| Limão | 55 | 1201 | 45 | 10 | 
| Manga | 1000 | 265 | 11 | 35 |

* Sensibilidade
* Especificidade
* Precisão
* Medida F1 
* Acurácia
* Coeficiente de Matthews
* AUC

Como você resolveria o problema dessa base de dados observando as métricas de avaliação?


In [15]:
#Criando uma classe Fruta com os parâmetros
#Verdadeiro Positivo, Verdadeiro Negativo, Falso Positivo, Falso Negativo

class Fruta:
  def __init__(self, vp, vn, fp, fn):
    self.vp = vp
    self.vn = vn
    self.fp = fp
    self.fn = fn

#Inicializando os objetos com seus valores 
maca = Fruta(70,1191,30,20)
laranja = Fruta(67,1157,33,54)
limao = Fruta(55,1201,45,10)
manga = Fruta(1000,265,11,35)

### Métricas de Desempenho:

#### Sensibilidade:

<img src="https://latex.codecogs.com/svg.image?sens(\hat{f})&space;=&space;rev(\hat{f})&space;=&space;TVP(\hat{f})&space;=&space;\frac{VP}{VP&plus;FN}" title="https://latex.codecogs.com/svg.image?sens(\hat{f}) = rev(\hat{f}) = TVP(\hat{f}) = \frac{VP}{VP+FN}" />


In [16]:
#Função para calcular a sensibilidade
def sens(VP, FN):
  sensibilidade = (VP/(VP+FN))
  return sensibilidade


In [17]:
#Sensibilidade de Maçã, Laranja, Limão e Manga respectivamente.
sens_maca = sens(maca.vp, maca.fn)
sens_laranja = sens(laranja.vp, laranja.fn)
sens_limao = sens(limao.vp, limao.fn)
sens_manga = sens(manga.vp, manga.fn)

#Printando as sensibilidades:
print('A sensibilidade de maçã:{:.2f}'.format(sens_maca))
print('A sensibilidade de laranja:{:.2f}'.format(sens_laranja))
print('A sensibilidade de limão:{:.2f}'.format(sens_limao))
print('A sensibilidade de manga:{:.2f}'.format(sens_manga))

#Printando a média de Sensibilidade:
sens_media = (sens_maca + sens_laranja + sens_limao + sens_manga ) / 4
print('\nA média de sensibilidade da Matriz de Confusão é:{:.2f}\n'.format(sens_media))

A sensibilidade de maçã:0.78
A sensibilidade de laranja:0.55
A sensibilidade de limão:0.85
A sensibilidade de manga:0.97

A média de sensibilidade da Matriz de Confusão é:0.79



#### Especificidade: 

<img src="https://latex.codecogs.com/svg.image?esp(\hat{f})&space;=&space;\frac{VN}{VN&plus;FP}&space;=&space;1-TFP(\hat{f})" title="https://latex.codecogs.com/svg.image?esp(\hat{f}) = \frac{VN}{VN+FP} = 1-TFP(\hat{f})" />

In [18]:
#Função para calcular a especificidade
def esp(VN, FP):
  especificidade = VN/(VN+FP)
  return especificidade

In [19]:
#Especificidade de Maçã, Laranja, Limão e Manga respectivamente.
espec_maca = esp(maca.vn, maca.fp)
espec_laranja = esp(laranja.vn, laranja.fp)
espec_limao = esp(limao.vn, limao.fp)
espec_manga = esp(manga.vn, manga.fp)

#Printando as especificidades
print('A especificidade de maçã:{:.2f}'.format(espec_maca))
print('A especificidade de laranja:{:.2f}'.format(espec_laranja))
print('A especificidade de limão:{:.2f}'.format(espec_limao))
print('A especificidade de manga:{:.2f}'.format(espec_manga))

#Printando a média de especificidade:
esp_media = (espec_maca + espec_laranja + espec_limao + espec_manga ) / 4
print('\nA média de especificidade da Matriz de Confusão é:{:.2f}\n'.format(esp_media))

A especificidade de maçã:0.98
A especificidade de laranja:0.97
A especificidade de limão:0.96
A especificidade de manga:0.96

A média de especificidade da Matriz de Confusão é:0.97



#### Precisão:

<img src="https://latex.codecogs.com/svg.image?prec(\hat{f})&space;=&space;\frac{VP}{VP&plus;FP}" title="https://latex.codecogs.com/svg.image?prec(\hat{f}) = \frac{VP}{VP+FP}" />


In [20]:
#Função para calcular a precisão
def prec(VP, FP):
  precisao = VP/(VP+FP)
  return precisao

In [21]:
#Precisão de Maçã, Laranja, Limão e Manga respectivamente.
prec_maca = prec(maca.vp, maca.fp)
prec_laranja = prec(laranja.vp, laranja.fp)
prec_limao = prec(limao.vp, limao.fp)
prec_manga = prec(manga.vp, manga.fp)

#Printando as precisões:
print('A precisão de maçã:{:.2f}'.format(prec_maca))
print('A precisão de laranja:{:.2f}'.format(prec_laranja))
print('A precisão de limão:{:.2f}'.format(prec_limao))
print('A precisão de manga:{:.2f}'.format(prec_manga))

#Printando a média de precisão:
prec_media = (prec_maca + prec_laranja + prec_limao + prec_manga ) / 4
print('\nA média de precisão da Matriz de Confusão é:{:.2f}\n'.format(prec_media))

A precisão de maçã:0.70
A precisão de laranja:0.67
A precisão de limão:0.55
A precisão de manga:0.99

A média de precisão da Matriz de Confusão é:0.73



#### Medida-F1:

<img src="https://latex.codecogs.com/svg.image?F&space;=&space;2&space;\cdot&space;\frac{prec&space;\cdot&space;sens}{prec&space;&plus;&space;sens}" title="https://latex.codecogs.com/svg.image?F = 2 \cdot \frac{prec \cdot sens}{prec + sens}" />

In [22]:
#Função para calcular a medida-F:
def medf(PREC, SENS):
  medidaF = 2*((PREC*SENS)/(PREC+SENS))
  return medidaF

In [23]:
#Medida-F de Maçã, Laranja, Limão e Manga respectivamente.
med_maca = medf(prec_maca, sens_maca)
med_laranja = medf(prec_laranja, sens_laranja)
med_limao = medf(prec_limao, sens_limao)
med_manga = medf(prec_manga, sens_manga)

#Printando as precisões:
print('A Medida-F1 de maçã:{:.2f}'.format(med_maca))
print('A Medida-F1 de laranja:{:.2f}'.format(med_laranja))
print('A Medida-F1 de limão:{:.2f}'.format(med_limao))
print('A Medida-F1 de manga:{:.2f}'.format(med_manga))

#Printando a média de Medida-F1:
med_media = (med_maca + med_laranja + med_limao + med_manga ) / 4
print('\nA média de Medida-F1 da Matriz de Confusão é:{:.2f}\n'.format(med_media))

A Medida-F1 de maçã:0.74
A Medida-F1 de laranja:0.61
A Medida-F1 de limão:0.67
A Medida-F1 de manga:0.98

A média de Medida-F1 da Matriz de Confusão é:0.75



#### Acurácia Total:

<img src="https://latex.codecogs.com/svg.image?ac(\hat{f})&space;=&space;\frac{VP&plus;VN}{n}" title="https://latex.codecogs.com/svg.image?ac(\hat{f}) = \frac{VP+VN}{n}" />

In [24]:
#Função para calcular a medida-F:
def ac(VP, VN, FP, FN):
  acuracia = (VP+VN)/(VP+VN+FP+FN)
  return acuracia

In [25]:
#Acurácia total de Maçã, Laranja, Limão e Manga respectivamente.
act_maca = ac(maca.vp, maca.vn, maca.fp, maca.fn)
act_laranja = ac(laranja.vp, laranja.vn, laranja.fp, laranja.fn)
act_limao = ac(limao.vp, limao.vn, limao.fp, limao.fn)
act_manga = ac(manga.vp, manga.vn, manga.fp, manga.fn)

#Printando as acurácias:
print('A acurácia total de maçã:{:.2f}'.format(act_maca))
print('A acurácia total de laranja:{:.2f}'.format(act_laranja))
print('A acurácia total de limão:{:.2f}'.format(act_limao))
print('A acurácia total de manga:{:.2f}'.format(act_manga))

#Printando a média de Acurácia total:
act_media = (act_maca + act_laranja + act_limao + act_manga ) / 4
print('\nA média de Acurácia Total da Matriz de Confusão é:{:.2f}\n'.format(act_media))

A acurácia total de maçã:0.96
A acurácia total de laranja:0.93
A acurácia total de limão:0.96
A acurácia total de manga:0.96

A média de Acurácia Total da Matriz de Confusão é:0.95



#### Coeficiente Matthews:

<img src="https://latex.codecogs.com/svg.image?phi&space;=&space;\frac{(VP\cdot&space;VN&space;-&space;FP\cdot&space;FN)}{\sqrt{(VP&plus;FP)\cdot&space;(VP&plus;FN)\cdot&space;(VN&plus;FP)\cdot&space;(VN&plus;FN)}}&space;" title="https://latex.codecogs.com/svg.image?phi = \frac{(VP\cdot VN - FP\cdot FN)}{\sqrt{(VP+FP)\cdot (VP+FN)\cdot (VN+FP)\cdot (VN+FN)}} " />


In [26]:
#Função para calcular o Coeficiente Matthews:
import math
def matthews(VP, VN, FP, FN):
  phi = ((VP*VN)-(FP*FN)) / math.sqrt(((VP+FP)*(VP+FN)*(VN+FP)*(VN+FN)))
  return phi

In [27]:
#Coeficiente Matthews de Maçã, Laranja, Limão e Manga respectivamente.
matthews_maca = matthews(maca.vp, maca.vn, maca.fp, maca.fn)
matthews_laranja = matthews(laranja.vp, laranja.vn, laranja.fp, laranja.fn)
matthews_limao = matthews(limao.vp, limao.vn, limao.fp, limao.fn)
matthews_manga = matthews(manga.vp, manga.vn, manga.fp, manga.fn)

#Printando a média de Coeficiente Matthews:
phi = (matthews_maca + matthews_laranja + matthews_limao + matthews_manga ) / 4
print('\nO Coeficiente Matthews da Matriz de Confusão é:{:.2f}\n'.format(phi))


O Coeficiente Matthews da Matriz de Confusão é:0.71



#### AUC:

In [28]:
import numpy as np

def auc(SENS, ESPEC):
  tvp = SENS
  tfp = (1-ESPEC)

  x = np.array([0, tvp, 1])
  y = np.array([0, tfp, 1])

  auc = np.trapz(x, y)
  return auc

In [29]:
#Printando a média de Coeficiente Matthews:
AUC = auc(sens_media, esp_media)
print('\nO AUC da Matriz de Confusão é:{:.2f}\n'.format(AUC))


O AUC da Matriz de Confusão é:0.88



#### Conclusão:

* Sensibilidade = 0.79
* Especificidade = 0.97
* Precisão = 0.73
* Medida F1 = 0.75
* Acurácia = 0.95
* Coeficiente de Matthews = 0.71
* AUC = 0.88

Para melhorar as métricas de avaliação, é necessário melhorar a fase de pré-processamento do banco de dados, portanto, executar uma limpeza na base de dados e re-treinar o modelo de classificação.

## Questão Extra: 

Elabore uma pergunta e responda sobre "Preconceito e Inteligência Artificial", refletindo aspectos importantes do método que uma pessoa engenheiro de dados/cientista de dados deve conhecer e se preocupar garantindo ética nos modelos criados.

### A Inteligência Artificial é capaz de expor preconceitos e atacar minorias?

A Inteligência Artificial serve para beneficiar a humanidade e está presente diariamente em nosso cotídiano, porém a tecnologia não está isenta de falhas. Uma base de dados mal desenvolvida pode ocasionar em modelos mal projetados para o mundo exterior, como por exemplo: Uma IA de classificação do Google com uma base de dados despreparada, ocasionou em um modelo onde indicou dois adolescentes negros como gorilas e, além disso, um chatbot criado pela Microsoft se tornou racista em 24h após sua ativação e exposição ao Twitter, rede onde o chatbot permaneceu mais ativo.

Entretanto, situações como essa não foram planejadas por seus desenvolvedores, mas ocasionada por eles. A Inteligência Artificial deve ter banco de dados bem planejados com a diversidade de seres humanos e medidas de segurança programadas por seus engenheiro/cientista de dados para evitar uma má aprendizagem de máquina.